GC job frequently fails due to failed delete request to GCS, should skip/retry instead?

We have noticed that very frequently a GC job will fail because either a single artifact or manifest delete request errored out (500 or 503) from the upstream storage backend (in our case GCS).

Would it be possible for the GC job to either skip or potentially retry the failed operation instead of failing the entire GC job? in the worst case this artifact or manifest will linger a while longer until the next GC run, but at least it won't fail the rest of the job.

For example:
```
2021-06-23T19:03:54Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:259]: delete the manifest with registry v2 API: <project>/<repo>, application/vnd.docker.distribution.manifest.v2+json, sha256:a1bbac70d2dc427b2ba7000656556dc1080e34fcbe82495274c98b30333721f5
                                                                                                                                                                                                                                                                                                                                                                                                                                                   
2021-06-23T19:04:32Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:262]: failed to delete manifest with v2 API, <project>/<repo>, sha256:a1bbac70d2dc427b2ba7000656556dc1080e34fcbe82495274c98b30333721f5, http status code: 500, body: {"errors":[{"code":"UNKNOWN","message":"unknown error","detail":{"DriverName":"gcs","Enclosed":{"Op":"Get","URL":"https://storage.googleapis.com/<gcs bucket>/docker/registry/v2/repositories/<project>/<repo>/_manifests/tags/c6be10d65c145417e34816a8535aab027404184a/current/link","Err":{"Op":"read","Net":"tcp","Source":{"IP":"172.16.6.30","Port":36382,"Zone":""},"Addr":{"IP":"x.x.x.x","Port":443,"Zone":""},"Err":{"Syscall":"read","Err":104}}}}}]}
 
2021-06-23T19:04:32Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:165]: failed to execute GC job at sweep phase, error: failed to delete manifest with v2 API: <project>/<repo>, sha256:a1bbac70d2dc427b2ba7000656556dc1080e34fcbe82495274c98b30333721f5: http status code: 500, body: {"errors":[{"code":"UNKNOWN","message":"unknown error","detail":{"DriverName":"gcs","Enclosed":{"Op":"Get","URL":"https://storage.googleapis.com/<gcs bucket>/docker/registry/v2/repositories/<project>/<repo>/_manifests/tags/c6be10d65c145417e34816a8535aab027404184a/current/link","Err":{"Op":"read","Net":"tcp","Source":{"IP":"172.16.6.30","Port":36382,"Zone":""},"Addr":{"IP":"x.x.x.x","Port":443,"Zone":""},"Err":{"Syscall":"read","Err":104}}}}}]}
```

--- a/src/jobservice/job/impl/gc/garbage_collection.go
+++ b/src/jobservice/job/impl/gc/garbage_collection.go
@@ -15,6 +15,7 @@
 package gc
 
 import (
+	"github.com/goharbor/harbor/src/lib"
 	"os"
 	"time"
 
@@ -270,9 +271,13 @@ func (gc *GarbageCollector) sweep(ctx job.Context) error {
 				}
 				// for manifest, it has to delete the revisions folder of each repository
 				gc.logger.Infof("delete manifest from storage: %s", blob.Digest)
-				if err := ignoreNotFound(func() error {
-					return gc.registryCtlClient.DeleteManifest(art.RepositoryName, blob.Digest)
-				}); err != nil {
+				if err := lib.RetryUntil(func() error {
+					return ignoreNotFound(func() error {
+						return gc.registryCtlClient.DeleteManifest(art.RepositoryName, blob.Digest)
+					})
+				}, lib.RetryCallback(func(err error, sleep time.Duration) {
+					gc.logger.Infof("failed to exec DeleteManifest retry after %s : %v", sleep, err)
+				})); err != nil {
 					if err := ignoreNotFound(func() error {
 						return gc.markDeleteFailed(ctx, blob)
 					}); err != nil {
@@ -294,9 +299,13 @@ func (gc *GarbageCollector) sweep(ctx job.Context) error {
 		// for the foreign layer, as it's not stored in the storage, no need to call the delete api and count size, but still have to delete the DB record.
 		if !blob.IsForeignLayer() {
 			gc.logger.Infof("delete blob from storage: %s", blob.Digest)
-			if err := ignoreNotFound(func() error {
-				return gc.registryCtlClient.DeleteBlob(blob.Digest)
-			}); err != nil {
+			if err := lib.RetryUntil(func() error {
+				return ignoreNotFound(func() error {
+					return gc.registryCtlClient.DeleteBlob(blob.Digest)
+				})
+			}, lib.RetryCallback(func(err error, sleep time.Duration) {
+				gc.logger.Infof("failed to exec DeleteBlob retry after %s : %v", sleep, err)
+			})); err != nil {
 				if err := ignoreNotFound(func() error {
 					return gc.markDeleteFailed(ctx, blob)
 				}); err != nil {

--- a/src/jobservice/job/impl/gc/util.go
+++ b/src/jobservice/job/impl/gc/util.go
@@ -2,9 +2,11 @@ package gc
 
 import (
 	"fmt"
+	"github.com/goharbor/harbor/src/lib"
 	"github.com/goharbor/harbor/src/lib/errors"
 	"github.com/goharbor/harbor/src/pkg/registry"
 	"github.com/gomodule/redigo/redis"
+	"time"
 )
 
 // delKeys ...
@@ -50,10 +52,11 @@ func v2DeleteManifest(repository, digest string) error {
 	if !exist {
 		return nil
 	}
-	if err := registry.Cli.DeleteManifest(repository, digest); err != nil {
-		return err
-	}
-	return nil
+	return lib.RetryUntil(func() error {
+		return registry.Cli.DeleteManifest(repository, digest)
+	}, lib.RetryCallback(func(err error, sleep time.Duration) {
+		fmt.Printf("failed to exec DeleteManifest retry after %s : %v\n", sleep, err)
+	}))
 }
 
 // ignoreNotFound ignores the NotFoundErr error

